<a href="https://colab.research.google.com/github/jakelaporte/MathematicalModeling/blob/master/Lsn15_DiscreteDynamicalSystems.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
options(warn=-1)
library(pracma)
library(ma391laporte)

# Discrete Time Dynamical Systems
Before attempting to go through this lesson, watch the following video and pay special attention to the typical case of having *n* eigenvalues and *n* independent eigenvectors for an $A_{nxn}$ matrix and take notes so that you can ask questions if you do not understand something: [Diagonalization and Powers of A](https://ocw.mit.edu/courses/mathematics/18-06-linear-algebra-spring-2010/video-lectures/lecture-22-diagonalization-and-powers-of-a/). <br><br>
In the video, you should have the following solution technique for a linear, discrete time, homogeneous dynamical system:<br><br>
$$\vec{u}_{k+1}=A\vec{u}_k\hspace{5pt}\text{initial value: }\vec{u}_0$$<br>
$$\vec{u}_1 = A\vec{u}_0 $$<br>
$$\vec{u}_2 = A\vec{u}_1 = A A\vec{u}_0=A^2\vec{u}_0 $$<br>
$$\vec{u}_k = A^k\vec{u}_0$$ <br>
So the solution involves finding $A^k$. In the video, you should have seen that the way we "really" solve this is by finding the eigenvalues ($\lambda_1,\lambda_2,...\lambda_n$) and eigenvectors($\vec{v}_1,\vec{v}_2,...\vec{v}_n$). Find the linear combination of the *n* independent eigenvectors of $\vec{u}_0$<br><br>
$$\vec{u}_0=c_1\vec{v}_1+c_2\vec{v}_2+c_3\vec{v}_3+...+c_n\vec{v}_n$$<br>
$$A\vec{u}_0=c_1A\vec{v}_1+c_2A\vec{v}_2+c_3A\vec{v}_3+...+c_nA\vec{v}_n$$
$$=c_1\lambda_1\vec{v}_1+c_2\lambda_2\vec{v}_2+c_3\lambda_3\vec{v}_3+...+c_n\lambda_n\vec{v}_n$$<br>
$$AA\vec{u}_0=A^2\vec{u}_0=c_1\lambda_1A\vec{v}_1+c_2\lambda_2A\vec{v}_2+c_3\lambda_3A\vec{v}_3+...+c_n\lambda_nA\vec{v}_n$$
$$=c_1\lambda_1^2\vec{v}_1+c_2\lambda_2^2\vec{v}_2+c_3\lambda_3^2\vec{v}_3+...+c_n\lambda_n^2\vec{v}_n$$<br>
$$\begin{equation}A^k\vec{u}_0=c_1\lambda_1^k\vec{v}_1+c_2\lambda_2^k\vec{v}_2+c_3\lambda_3^k\vec{v}_3+...+c_n\lambda_n^k\vec{v}_n\end{equation}$$<br>
This solution answers the question of stability for the system. How? Notice that if the eigenvalues of the system are $|\lambda_i|$ are all less than one (this becomes the test for discrete systems), then $A^k\vec{u}_0$ goes to zero and the system is stable.

## Example (USMA text)
A simple model of a fish population with three age groups - age under 1 year (young, $y_n$), age between 1 and 2 years old (middle, $m_n$) and age over two years (adult, $a_n$). 
#### Modeling the Problem #####
Variables: <ol>
- $n$ = time (years) after lake is stocked<br>
- $y_n$ = the number (in thousands) of young fish <br> 
- $m_n$ = the number (in thousands) of middle aged fish<br>
- $a_n$ = the number (in thousands) of adult fish <br>
- $x_n$ = the vector of [young middle adult] fish in thousands - these are the state variables<br>
- $x_{n+1} = Ax_n$ where $A$ is found by the relationship between the state variables (see below)<br><br>
- $x_0 = \begin{bmatrix}2\\2\\0\end{bmatrix}$ 
</ol>

Assumptions:<ol>
- Each year 50% of the population in the first age group survive, 80% of the middle age group make it to adulthood, and 50% of adults continue being adults.
- On average 30% of young fish have offspring, 52% of middle-aged fish have offspring, and 28% of adult fish have offspring.
</ol>
These assumptions lead to the following equations: <br>
Young fish population-> $y_{n+1}=0.3y_n+0.52m_n+0.28a_n$.<br> 
Middle-age population -> $m_{n+1}=0.5y_n$<br>
Adult population -> $a_{n+1}=0.8m_n+0.5a_n$ <br><br>
Putting this equation into the form that we need to solve it:<br><br>
$$\vec{x}_{n+1}=\begin{bmatrix}y_{n+1}\\m_{n+1}\\a_{n+1}\end{bmatrix} = \begin{bmatrix}0.3y_n+0.52m_n+0.28a_n\\0.5y_n\\0.8m_n+0.5a_n \end{bmatrix}=\begin{bmatrix}0.3&0.52&0.28\\0.5&0&0\\0&0.8&0.5\end{bmatrix}\begin{bmatrix}y_n\\m_n\\a_n\end{bmatrix}=A\vec{x}_n$$<br><br>
The question here is what are the dynamics in this problem? Does the population grow forever or does it die out? We use the theory from the video / and in the last section to understand what is happening. We could just run this with [2,2,0] and see what happens, but we would rather understand what is happening based on the math.

In [0]:
print("A")
A = matrix(c(.3,.52,.28,.5,0,0,0,.8,.5),nrow=3,byrow=T);print(A)
print("u0")
u0 = matrix(c(2,2,0));print(u0)
ev = eigen(A)
print("Lambda-diagonal matrix with eigenvalues in diagonal")
Lambda = diag(ev$values);print(Lambda)
lambda = ev$values
print("S")
S=ev$vectors;print(S)

[1] "A"
     [,1] [,2] [,3]
[1,]  0.3 0.52 0.28
[2,]  0.5 0.00 0.00
[3,]  0.0 0.80 0.50
[1] "u0"
     [,1]
[1,]    2
[2,]    2
[3,]    0
[1] "Lambda-diagonal matrix with eigenvalues in diagonal"
     [,1] [,2] [,3]
[1,]  0.9  0.0  0.0
[2,]  0.0 -0.2  0.0
[3,]  0.0  0.0  0.1
[1] "S"
           [,1]       [,2]        [,3]
[1,] -0.6270597  0.2547139 -0.08908708
[2,] -0.3483665 -0.6367848 -0.44543540
[3,] -0.6967330  0.7277540  0.89087081


Using the matrix of normalized eigenvectors from above, finding the linear combinations is simply solving the following system of equations:<br><br>
$$S \vec{c}=\vec{u}_0$$<br>
where $S$ represents the eigenvector matrix, $vec{c}$ is the vector of linear combinations of the eigenvectors needed to sum up to the initial starting value $\vec{u}_0$.

In [0]:
c = solve(S,u0);print(c)

              [,1]
[1,] -2.870540e+00
[2,]  3.106569e-15
[3,] -2.244994e+00


The analytic solution to the system of equations (we will verify this below) is the following vector equation where we only need to substitute in the value of $k$ that we are interested in (note that we have used the notation from the problem $n$ instead of $k$):<br><br>
$$A^n\vec{u}_0=\vec{u}(n)=(-2.87)(0.9)^n\begin{bmatrix}-0.627097\\-0.3483665\\-0.6967330 \end{bmatrix}+(0)(-0.2)^n\begin{bmatrix}0.2547139\\-0.6367848\\0.7277540 \end{bmatrix}+(-2.245)(0.1)^n\begin{bmatrix}-0.08908708\\-0.44543540\\0.89087081 \end{bmatrix}$$

In [0]:
## Verify that the matrix multiplication gives the same result as the analytic vector solution ##
un = function(n){round(c[1]*lambda[1]^n*S[,1]+c[3]*lambda[3]^n*S[,3],3)}
u = u0
for (i in 0:10){
    print("=======================")
    print(i)
    print("un = eigenvalues")
    print(un(i))
    ## Compare with matrix multiplication uk = A^k*u0##
    print("A^n*u0")
    print(t(round(u,3)))
    u = A%*%u
}

[1] "======================="
[1] 0
[1] "un = eigenvalues"
[1] 2 2 0
[1] "A^n*u0"
     [,1] [,2] [,3]
[1,]    2    2    0
[1] "======================="
[1] 1
[1] "un = eigenvalues"
[1] 1.64 1.00 1.60
[1] "A^n*u0"
     [,1] [,2] [,3]
[1,] 1.64    1  1.6
[1] "======================="
[1] 2
[1] "un = eigenvalues"
[1] 1.46 0.82 1.60
[1] "A^n*u0"
     [,1] [,2] [,3]
[1,] 1.46 0.82  1.6
[1] "======================="
[1] 3
[1] "un = eigenvalues"
[1] 1.312 0.730 1.456
[1] "A^n*u0"
      [,1] [,2]  [,3]
[1,] 1.312 0.73 1.456
[1] "======================="
[1] 4
[1] "un = eigenvalues"
[1] 1.181 0.656 1.312
[1] "A^n*u0"
      [,1]  [,2]  [,3]
[1,] 1.181 0.656 1.312
[1] "======================="
[1] 5
[1] "un = eigenvalues"
[1] 1.063 0.590 1.181
[1] "A^n*u0"
      [,1] [,2]  [,3]
[1,] 1.063 0.59 1.181
[1] "======================="
[1] 6
[1] "un = eigenvalues"
[1] 0.957 0.531 1.063
[1] "A^n*u0"
      [,1]  [,2]  [,3]
[1,] 0.957 0.531 1.063
[1] "======================="
[1] 7
[1] "un = eigenvalues"
[

#### Conclusions ####
Since the analytic solution to the matrix multiplication can give us insight into the dynamics of the problem, we use that version and know that when all of the eigenvalues are less than one in absolute value $|\lambda_i|$ for all $i$, then the equilibrium solution is stable.

Finally, understand that the analysis done above is with respect to the recursion equations and we can very easily turn this back into a difference equation problem and use the path function from previous lessons. We need to understand how to convert from one to the other, so follow the simple logic.<br><br>
$$ \begin{bmatrix}y_{n+1}\\m_{n+1}\\a_{n+1} \end{bmatrix}-\begin{bmatrix}y_{n}\\m_{n}\\a_{n}\end{bmatrix} = \begin{bmatrix}\Delta y \\ \Delta m \\ \Delta a\end{bmatrix} = \begin{bmatrix}0.3y_n+0.52m_n+0.28a_n\\0.5y_n\\0.8m_n+0.5a_n \end{bmatrix}-\begin{bmatrix}y_{n}\\m_{n}\\a_{n}\end{bmatrix}=\begin{bmatrix}-0.7y_n+0.52m_n+0.28a_n\\ 0.5y_n-m_n\\0.8m_n-0.5a_n \end{bmatrix}$$ <br><br>
In terms of the state variables: <br>
$$\begin{bmatrix}\Delta x_1 \\ \Delta x_2 \\ \Delta x_3\end{bmatrix} =\begin{bmatrix}-0.7x_1+0.52x_2+0.28x_3\\ 0.5x_1-x_2\\ 0.8x_2-0.5x_3 \end{bmatrix} $$<br><br>
As you can see below, we use the path function with these difference equations to get the same results.

In [0]:
## Replace this path function with the one that we used previously (it assumed a 2D system)#####
path = function(f,x0,deltat=0.01,N=1000,tol=1E-4){
  len = length(x0)
  points=matrix(0,ncol=len)
  points[1,] = x0
  n = 0
  p = c(1,1)
  while(norm(p)>tol & n<N){
    n=n+1
    p = f(x0)*deltat
    x0=x0+p
    points = rbind(points,x0)
  }

  rownames(points)=0:n
  return(points)
}

In [0]:
## Compare these results with the recursion equations above ##
f = function(x){c(-0.7*x[1]+0.52*x[2]+0.28*x[3],
                 0.5*x[1]-x[2],
                 0.8*x[2]-.5*x[3])}
x0=c(2,2,0)
path(f,x0,N=10,deltat=1)

[,1]      [,2]      [,3]     
0  2.0000000 2.0000000 0.0000000
1  1.6400000 1.0000000 1.6000000
2  1.4600000 0.8200000 1.6000000
3  1.3124000 0.7300000 1.4560000
4  1.1810000 0.6562000 1.3120000
5  1.0628840 0.5905000 1.1809600
6  0.9565940 0.5314420 1.0628800
7  0.8609344 0.4782970 0.9565936
8  0.7748410 0.4304672 0.8609344
9  0.6973569 0.3874205 0.7748410
10 0.6276212 0.3486784 0.6973569